File: 39. Combination Sum/39. Combination Sum.dart
class Solution {
  List<List<int>> combinationSum(List<int> candidates, int target) {
    candidates.sort();
    List<List<int>> result = [];
    backtrack(candidates, target, 0, [], result);
    return result;
  }
  
  void backtrack(List<int> candidates, int target, int index, List<int> current, List<List<int>> result) {
    if (target == 0) {
      result.add(List.from(current));
      return;
    }
    
    for (int i = index; i < candidates.length; i++) {
      if (candidates[i] > target) break;
      current.add(candidates[i]);
      backtrack(candidates, target - candidates[i], i, current, result);
      current.removeLast();
    }
  }
}

File: 39. Combination Sum/39. Combination Sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        Array.Sort(candidates);
        var result = new List<IList<int>>();
        Backtrack(candidates, target, 0, new List<int>(), result);
        return result;
    }
    
    private void Backtrack(int[] candidates, int target, int index, List<int> current, List<IList<int>> result) {
        if (target == 0) {
            result.Add(new List<int>(current));
            return;
        }
        
        for (int i = index; i < candidates.Length; i++) {
            if (candidates[i] > target) break;
            current.Add(candidates[i]);
            Backtrack(candidates, target - candidates[i], i, current, result);
            current.RemoveAt(current.Count - 1);
        }
    }
}